server/svix-server/src/v1/endpoints/endpoint/crud.rs

```diff
@@ -47,156 +47,172 @@ pub(super) async fn list_endpoints(
+    app: application::Model,
+    mut data: EndpointIn,
+) -> Result<(endpoint::Model, endpointmetadata::Model)> {
+    let key = data.key_take_or_generate(&cfg.encryption, &cfg.default_signature_type)?;
+
+    let mut endp = endpoint::ActiveModel::new(app.id, key);
+    let metadata =
+        endpointmetadata::ActiveModel::new(endp.id.clone().unwrap(), mem::take(&mut data.metadata));
+    data.update_model(&mut endp);
+
+    let (endp, metadata) = {
+        let txn = ctx!(db.begin().await)?;
+        let endp = ctx!(endp.insert(&txn).await)?;
+        let metadata = ctx!(metadata.upsert_or_delete(&txn).await)?;
+        ctx!(txn.commit().await)?;
+        (endp, metadata)
+    };
+
+    op_webhooks
+        .send_operational_webhook(
+            &app.org_id,
+            OperationalWebhook::EndpointCreated(EndpointEvent::new(app.uid.as_ref(), &endp)),
+        )
+        .await?;
+
+    Ok((endp, metadata))
+}
+
 pub(super) async fn create_endpoint(
     Extension(ref db): Extension<DatabaseConnection>,
     Extension(ref cfg): Extension<Configuration>,
     Extension(op_webhooks): Extension<OperationalWebhookSender>,
     permissions::Application { app }: permissions::Application,
-    ValidatedJson(mut data): ValidatedJson<EndpointIn>,
+    ValidatedJson(data): ValidatedJson<EndpointIn>,
 ) -> Result<(StatusCode, Json<EndpointOut>)> {
     if let Some(ref event_types_ids) = data.event_types_ids {
         validate_event_types(db, event_types_ids, &app.org_id).await?;
     }
     validate_endpoint_url(&data.url, cfg.endpoint_https_only)?;
 
-    let endp = endpoint::ActiveModel {
-        app_id: Set(app.id),
-        key: Set(data.key_take_or_generate(&cfg.encryption, &cfg.default_signature_type)?),
-        ..data.into()
-    };
-
-    let ret = ctx!(endp.insert(db).await)?;
+    let (endp, metadata) = ctx!(create_endp_data(db, cfg, &op_webhooks, app, data).await)?;
 
-    op_webhooks
-        .send_operational_webhook(
-            &app.org_id,
-            OperationalWebhook::EndpointCreated(EndpointEvent::new(app.uid.as_ref(), &ret)),
-        )
-        .await?;
-
-    Ok((StatusCode::CREATED, Json(ret.into())))
+    Ok((StatusCode::CREATED, Json((endp, metadata.data).into())))
 }
 
 pub(super) async fn get_endpoint(
     Extension(ref db): Extension<DatabaseConnection>,
     Path((_app_id, endp_id)): Path<(ApplicationIdOrUid, EndpointIdOrUid)>,
     permissions::Application { app }: permissions::Application,
 ) -> Result<Json<EndpointOut>> {
-    let endp = ctx!(
+    let (endp, metadata) = ctx!(
         endpoint::Entity::secure_find_by_id_or_uid(app.id, endp_id)
+            .find_also_related(endpointmetadata::Entity)
             .one(db)
             .await
     )?
     .ok_or_else(|| HttpError::not_found(None, None))?;
-    Ok(Json(endp.into()))
+
+    let metadata = metadata.map(|m| m.data).unwrap_or_default();
+
+    Ok(Json((endp, metadata).into()))
+}
+
+async fn update_endp_data(
```

Comment:
```suggestion
async fn update_endp_from_data(
```
I guess?

Same for `create`